book/src/what_is_chalk/walkthrough.md

```diff
@@ -0,0 +1,218 @@
+  "if" and `conditions = cond1 && cond2 && ...`
+* A universally quantified clause of the form
+  `forall<T> { consequence :- conditions }`
+  * `forall<T> { ... }` is used to represent [universal quantification]. See the
+    section on [Lowering to logic][lowering-forall] for more information.
+  * A key thing to note about `forall` is that we don't allow you to "quantify"
+    over traits, only types and regions (lifetimes). That is, you can't make a
+    rule like `forall<Trait> { u32: Trait }` which would say "`u32` implements
+    all traits". You can however say `forall<T> { T: Trait }` meaning "`Trait`
+    is implemented by all types".
+  * `forall<T> { ... }` is represented in the code using the [`Binders<T>`
+    struct][binders-struct].
+
+*See also: [Goals and Clauses][goals-and-clauses]*
+
+This is where we encode the rules of the trait system into logic. For
+example, if we have the following Rust:
+
+```rust,ignore
+impl<T: Clone> Clone for Vec<T> {}
+```
+
+We generate the following program clause:
+
+```rust,ignore
+forall<T> { (Vec<T>: Clone) :- (T: Clone) }
+```
+
+This rule dictates that `Vec<T>: Clone` is only satisfied if `T: Clone` is also
+satisfied (i.e. "provable").
+
+Similar to [`chalk::program::Program`][chalk-program] which has "rust-like
+things", chalk_ir defines [`ProgramEnvironment`] which is "pure logic".
+The main field in that struct is `program_clauses`, which contains the
+[`ProgramClause`]s generated by the rules module.
+
+### Rules ([chalk_solve])
+
+The `chalk_solve` crate ([source code][chalk_solve]) defines the logic rules we
+use for each item in the Rust IR. It works by iterating over every trait, impl,
+etc. and emitting the rules that come from each one.
+
+*See also: [Lowering Rules][lowering-rules]*
+
+#### Well-formedness checks
+
+As part of lowering to logic, we also do some "well formedness" checks. See
+the [`chalk_solve::wf` source code][solve-wf-src] for where those are done.
+
+*See also: [Well-formedness checking][wf-checking]*
+
+#### Coherence
+
+The method `CoherenceSolver::specialization_priorities` in the `coherence` module
+([source code][coherence-src]) checks "coherence", which means that it
+ensures that two impls of the same trait for the same type cannot exist.
+
+### Solver ([chalk_solve])
+
+Finally, when we've collected all the program clauses we care about, we want
+to perform queries on it. The component that finds the answer to these
+queries is called the *solver*.
+
+*See also: [The SLG Solver][slg]*
+
+## Crates
+
+Chalk's functionality is broken up into the following crates:
+- [**chalk_engine**][chalk_engine]: Defines the core [SLG solver][slg].
+- [**chalk_rust_ir**][chalk_rust_ir], containing the "HIR-like" form of the AST
+- [**chalk_ir**][chalk_ir]: Defines chalk's internal representation of
+  types, lifetimes, and goals.
+- [**chalk_solve**][chalk_solve]: Combines `chalk_ir` and `chalk_engine`,
```

Comment:
We should open a followup issue to make sure we document where the recursive solver fits into all of this. I don't think it needs to go in this PR, since we are considering splitting it out of `chalk-solve` anyways.